src/main/scala/org/broadinstitute/dsde/workbench/sam/backgroundProcess/CryptominingSubscriber.scala

```diff
@@ -0,0 +1,47 @@
+package org.broadinstitute.dsde.workbench.sam
+package backgroundProcess
+
+import cats.effect.{ContextShift, IO}
+import fs2.Stream
+import io.chrisdavenport.log4cats.StructuredLogger
+import io.circe.Decoder
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import org.broadinstitute.dsde.workbench.model.WorkbenchUserId
+import org.broadinstitute.dsde.workbench.sam.service.UserService
+import org.broadinstitute.dsde.workbench.sam.util.SamRequestContext
+
+// Each instance of Sam will run a subscriber, since they all subscribe to the same topic via same subscription,
+// messages will be distributed among subscribers instead of distributed to all of them, which is what we want.
+class CryptominingSubscriber(subscriber: GoogleSubscriber[IO, CryptominingUserMessage],
+                             userService: UserService
+                            )(implicit logger: StructuredLogger[IO], cs: ContextShift[IO]) {
+  val process: Stream[IO, Unit] = subscriber.messages
+    .parEvalMapUnordered(2)(messageHandler) // concurrency is low since we don't need high performance in disabling cryptomining users.
+    .handleErrorWith(error => Stream.eval(logger.error(error)("Failed to initialize message processor")))
+
+  private def messageHandler(event: Event[CryptominingUserMessage]): IO[Unit] = {
+    for {
+      res <- messageResponder(event.msg).attempt
+      _ <- res match {
+        case Left(e)  => logger.error(e)("Fail to process pubsub message")
+        case Right(_) => IO.unit
+      }
+      // acking the message whether or not we successfully processed the message since this is not essential part of
+      // terra functionality. That is, if we didn't disable a bad user, it's no big deal.
+      // To conditionally acking message, we'll need to config a dead letter topic where messages reach max retry can go to,
+      // which we're not doing for this use case.
```

Comment:
I disagree with this. Maybe for cryptomining it is not a big deal but I think the general contract for sam should be stronger. I think we should only ack on success.

Reply:
this ties back to https://github.com/broadinstitute/sam/pull/494/files/75635f99381bdba2035a10e72098ac34e11fd052#r552697473...

The reason why I'm choosing to ack blindly here is because if we choose to nack, then we should set up max retry, which means we should set up a deadlettertopic for receiving messages that have exceeded max retry, otherwise, nacked messages will be retried forever (I'm not certain if it's forever or there's a default max at some point, but from what we see in leo, it looked like forever). I didn't want to introduce more complexity if not necessary, but if you think it's worth it, I'm happy to add those